python/pynq/_pynq/_video/_framebuffer.c

```diff
@@ -0,0 +1,321 @@
+    self = (framebufferObject *)type->tp_alloc(type, 0);
+    return (PyObject *)self;
+}
+
+/*
+ * __init()__ method
+ *
+ * Python Constructor: frame([single_frame])
+ * set single_frame to 1 if you want this object to hold a single frame
+ * (that will be available at index 0)
+ */
+static int framebuffer_init(framebufferObject *self, PyObject *args)
+{
+    self->width = 0;
+    self->height = 0;
+    self->color_depth = DEFAULT_COLOR_DEPTH;
+    self->size = 0;
+
+    if (!PyArg_ParseTuple(args, "II|I", &self->width, &self->height, &self->color_depth))
+        return -1;
+
+    self->size = self->width * self->height * self->color_depth;
+    if ((self->frame_buffer = (u8 *)cma_alloc(sizeof(u8) * self->size, 0)) == NULL) {
+        PyErr_Format(PyExc_MemoryError,"Unable to allocate memory");
+        return -1;
+
+    }
+    return 0;
+}
+
+/*
+ * exposing members:
+ *  width: Width of the image
+ *  height: Height of the image
+ *  color_depth: Number of bytes a single pixel required (normally 3 for
+ *      RGB)
+ *  size: Total number of bytes the framebuffer holds
+ */
+static PyMemberDef framebuffer_members[] =
+{
+    {"width", T_UINT, offsetof(framebufferObject, width),READONLY,
+     "width of the image"},
+    {"height", T_UINT, offsetof(framebufferObject, height),READONLY,
+     "height of the image"},
+    {"color_depth", T_UINT, offsetof(framebufferObject, color_depth),READONLY,
+     "number of bytes per pixel"},
+    {"size", T_UINT, offsetof(framebufferObject, size),READONLY,
+     "total size of the image (in bytes) = width * height * color_depth"},
+    {NULL}  /* Sentinel */
+};
+
+/*****************************************************************************/
+/* functions used to set or get a frame or get (physical) frame address      */
+/* used by the frame member functions and by _capture.c and _display.c       */
+
+/*
+ * get a physical frame address as long
+ *
+ * The physical address is accessable by the FPGA. Giving the normal address
+ * of the framebuffer (E.G. &self->framebuffer[0]) would cause the kernel to
+ * freeze because that address is virtual (only relevant for the user
+ * application). The kernel does some address mangling in order to allow
+ * users to work with comfortable address spaces like 0x00000000 but if that
+ * virtual address was passed to the kernel the FPGA may try and access
+ * memory at an illegal address space. If so there is no 'segfault' that would
+ * protect the kernel, instead the kernel may actually write to that space and
+ * cause the kernel to be unstable.
+ *
+ *
+ */
+PyObject *get_framebuffer_phyaddr(framebufferObject *self)
+{
+    unsigned int ret = cma_get_phy_addr(self->frame_buffer);
+    return PyLong_FromUnsignedLong(ret);
+}
+
+/*
+ * get a bytearray of the framebuffer
+ */
+PyObject *get_framebuffer(framebufferObject *self)
+{
+    return PyByteArray_FromStringAndSize((char *)self->frame_buffer, self->width * self->height * self->color_depth);
```

Comment:
Is it possible to break this long line to 2, in order to make sure all lines do not exceed 79 characters?